[lldb] Removed gdbserver ports map from lldb-server

slydiman opened the pull request:
Listen to gdbserver-port, accept the connection and run `lldb-server gdbserver --fd` on all platforms. 
Parameters --min-gdbserver-port and --max-gdbserver-port are deprecated now.

This is the part 2 of #101283.

Fixes #97537.

DavidSpickett commented:
Thanks for putting so much work into this. I will test this out with my VM setup.

> Parameters --min-gdbserver-port and --max-gdbserver-port are deprecated now.

Are they just consumed but the port number is not used because it can use the first port for everything?

(I do see the logic of leaving the options in at first, so we can tell whether any problems are from the changes in this PR or purely from removing the options)

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -601,15 +570,24 @@ int main_platform(int argc, char *argv[]) {
     client_handle(platform, inferior_arguments);
   } while (g_server);
 
+  // FIXME: Make TCPSocket::CloseListenSockets() public and implement
+  // Acceptor::Close().
+  /*
+  if (acceptor_gdb && gdb_thread.IsJoinable()) {
+    g_listen_gdb = false;
+    static_cast<TCPSocket *>(acceptor_gdb->m_listener_socket_up.get())
+        ->CloseListenSockets();
+    lldb::thread_result_t result;
+    gdb_thread.Join(&result);
+  }
+  */
+
```
This won't work. This is exactly the pattern from that kernel bug we were arguing over. It's just not possible to safely close an fd while another thread may be operating on it.

You need to use an out-of-band mechanism (e.g. a pipe) to notify the thread that it's time to terminate. However, since that will involve some form of multiplexing (e.g. `select(2)`). I would strongly prefer to just multiplex over the two sockets we are accepting. We should be able to use the `MainLoop` class for that -- just register the two sockets with the class, and let it call the right callback when needed. Basically, the code should look similar to [this](https://github.com/llvm/llvm-project/blob/f71b63865140cf3c286baf3a77ba3e467f929504/lldb/source/Host/common/TCPSocket.cpp#L267), except that it will be listening to two unrelated sockets.

slydiman replied:
Oh, Linux... Such pattern works just fine on Windows. I saw MainLoop. But I think it is much safer just exit the app. Terminating the main thread is enough. I will remove this comment in the code and the g_listen_gdb flag.

labath replied:
<rant>It's not just linux, that kind of pattern is racy on all posix operating systems. It's possible that other OSs will not hang in this situation, but that doesn't mean that usage is safe. Even on windows, that can cause you to call (WSA)accept on a dangling socket descriptor -- since you can never be sure whether the other thread has already called the function, or is it merely about to do that. It's less likely to blow up in your face because windows does not recycle FDs like posix systems do, but I wouldn't recommend that anywhere.</rant>

Exiting the main thread may be convenient, but I definitely wouldn't call it "safe", as that means the other thread may be running, executing random code and memory, while the main thread is shutting everything down and deleting that memory. It may not matter often, as the process is done anyway, but if anyone was observing the exit code, he may be suprised to see a SEGV instead of a clean exit.

Reasoning about a singlethreaded application is much easier than for a multithreaded one. We should already have all of the pieces to do this the right way, so let's just do that.

labath commented on lldb/unittests/tools/lldb-server/tests/TestClient.h:
```diff
@@ -20,6 +20,10 @@
 #include <optional>
 #include <string>
 
+#ifdef SendMessage
+#undef SendMessage
+#endif
+
```
I guess that means we're including a windows header somewhere. Any chance to avoid that? Llvm is pretty strict about that. Lldb is not, though I think it's a good practice.

slydiman replied:
Note TestClient.cpp and LLGSTest.cpp already contain this code. Windows headers are necessary for socket definitions. It is impossible to separate them. This fix is necessary because the order of headers has been changed. Note we can remove the same code in TestClient.cpp and LLGSTest.cpp. I just minimized changes. Should I remove it from TestClient.cpp and LLGSTest.cpp?

slydiman replied:
I will retest it after #104439. Probably we won't need it.

labath replied:
If it's just due to header moves that's fine.

It's definitely possible to isolate the windows headers. Lllvm does that by making sure it only includes the system headers in implementation files (of the library wrapping them). It'd be nice to have lldb do the same thing, but we're pretty far off.

labath commented on lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp:
```diff
@@ -666,7 +756,23 @@ ConnectionStatus
-#if LLDB_ENABLE_POSIX
+#ifdef _WIN32
+  int64_t fd = -1;
+  if (!s.getAsInteger(0, fd)) {
+    // Assume we own fd.
+    std::unique_ptr<TCPSocket> tcp_socket =
+        std::make_unique<TCPSocket>((NativeSocket)fd, true, false);
+    m_io_sp = std::move(tcp_socket);
+    m_uri = s.str();
+    return eConnectionStatusSuccess;
+  }
```
I don't exactly have an alternative in mind yet, but I'm not sure this is a good idea. File descriptors are a thing on windows as well (though, IIUC, they're more of like an in-process thing rather than a kernel entity), and we do have code which treats them as such, so I think it could be confusing that this functions considers an "fd" to be a socket handle.

slydiman replied:
Right, file descriptors on Windows are in-process CRT specific things. But we already have this API. I don't see any point in complicating things here.

labath replied:
I do. I don't think it's a good state if half of our "fd" refer to file descriptors and others to handles. (In reality, it's not half/half, but more like 90/10, and this is the minority case/)

slydiman replied:
What are you suggesting here? We are trying to reduce usage of `#ifdef _WIN32`. HANDLE(s) are Windows specific. We can refactor it later if necessary. I'm trying to reduce changes.

slydiman replied:
ConnectionFileDescriptorPosix.cpp is unchanged now.
I have moved the SharedSocket using to lldb-gdbserver.cpp.
I don't like the idea about SharedSocket::GetSocket() because of ambiguity Socket/TCPSocket/UDPSocket and additional parameters should_close, child_processes_inherit, etc. It may reduce using of NativeSocket, but will not reduce the code in general.

labath replied:
Okay, I think I can live with that.

labath reviewed (commented):
The termination part definitely needs to be done differently, but overall, I think this is a pretty good start.

(PSA: I will be OOO all of next week.)

slydiman commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunication.cpp:
```diff
@@ -879,20 +879,12 @@ lldb::thread_result_t GDBRemoteCommunication::ListenThread() {
-Status GDBRemoteCommunication::StartDebugserverProcess(
-    const char *url, Platform *platform, ProcessLaunchInfo &launch_info,
-    uint16_t *port, const Args *inferior_args, int pass_comm_fd) {
+bool GDBRemoteCommunication::GetDebugserverPath(
+    Platform *platform, FileSpec &debugserver_file_spec) {
```
I tried to minimize changes here. `bool debugserver_exists` already was here. GetDebugserverPath() is used in GDBRemoteCommunicationServerPlatform.cpp for checking debugserver w/o launching. Invalid FileSpec is not convenient for this purpose.

slydiman replied:
Updated.

slydiman commented:
> Are they just consumed but the port number is not used because it can use the first port for everything?
> 
> (I do see the logic of leaving the options in at first, so we can tell whether any problems are from the changes in this PR or purely from removing the options)

We need only 1 port defined by `--gdbserver-port`. I kept `--min-gdbserver-port` and `--max-gdbserver-port` for backward compatibility. 
```
lldb-server p --server --listen *:1234 --gdbserver-port 1235 --min-gdbserver-port 1236 --max-gdbserver-port 1237
```
1235 is used for gdb
```
lldb-server p --server --listen *:1234 --min-gdbserver-port 1236 --max-gdbserver-port 1237
```
1236 is used for gdb
```
lldb-server p --server --listen *:1234 --max-gdbserver-port 1237 --min-gdbserver-port 1236
```
1237 is used for gdb

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -47,111 +48,10 @@ using namespace llvm;
 static int g_debug = 0;
 static int g_verbose = 0;
 static int g_server = 0;
+static volatile bool g_listen_gdb = true;
```
yeah, but volatile still isn't the right choice for that. It would have to be an atomic var.

DavidSpickett commented:
Only had time to build so far but:
```
/home/davspi01/work/open_source/llvm-project/lldb/tools/lldb-server/lldb-platform.cpp: In function ‘int main_platform(int, char**)’:
/home/davspi01/work/open_source/llvm-project/lldb/tools/lldb-server/lldb-platform.cpp:354:49: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  354 |       if (port_offset < LOW_PORT || port_offset > HIGH_PORT) {
      |                                                 ^
/home/davspi01/work/open_source/llvm-project/lldb/tools/lldb-server/lldb-platform.cpp:372:41: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  372 |       if (portnum < LOW_PORT || portnum > HIGH_PORT) {
      |                                         ^
[2796/2796] Linking CXX executable bin/lldb-server
```
I think this warning was already there, I'm just seeing it now in a clean build. `u` means `unsigned` and `portnum` is `uint16_t`.

slydiman commented:
> I think this warning was already there, I'm just seeing it now in a clean build. `u` means `unsigned` and `portnum` is `uint16_t`.

I will revert the original definition `#define HIGH_PORT (49151u)`. Thanks.

DavidSpickett commented on lldb/docs/man/lldb-server.rst:
```diff
@@ -147,15 +147,8 @@ GDB-SERVER CONNECTIONS
- Specify the range of ports that can be used for gdb-server connections. Both
- options need to be specified simultaneously. Overrides --gdbserver-port.
+ Define a port to be used for gdb-server connections. This port is used for
+ multiple connections.
```
I don't know much about how ports work, can you explain briefly how that single port knows to route traffic from each client lldb to the right lldb-server gdbserver?

It appears to "just work" but I'm curious.

slydiman replied:
Before: 
- `lldb-server platform`
  - got the query `qLaunchGDBServer`
  - request a gdb port from the map (it may be 0)
  - launch `lldb-server gdbserver host:port`
  - if the port was 0, read the actual port back via pipe
  - answer OK and report the actual port and pid
- lldb client connects to gdbserver using host:port (if handshake failed, lldb client kills gdbserver by pid via platform protocol)

After:
- `lldb-server platform`
  - listen to gdbserver-port
  - got the query `qLaunchGDBServer`
  - answer OK and report always known gdbserver-port and ibvalid pid
- lldb client connects to gdbserver using host:port (do nothing if handshake failed)
- `lldb-server platform` got connection and launch `lldb-server gdbserver --fd socket`
  - `lldb-server platform` will transfer the socket info to gdbserver via pipe on Windows (or kill gdbserver if it failed)

DavidSpickett replied:
Thanks, I understood that bit.

What I was missing was basic socket theory, that on the remote side we can look at the IP address of the sender so IP A to port N can be differentiated from IP B sending to port N. It's been a very long time since I had to think about this, but in hindsight, of course lots of things multiplex this way.

Nothing wrong with the code, if anything it was a bit too magical :) And a good improvement for the way I use lldb in my work.

DavidSpickett commented:
I did some testing with my usual setup which is x86 Linux to AArch64 Linux running in a simulator.

```
<simulator>
$ ./lldb-server platform --server --listen 0.0.0.0:1234 --gdbserver-port 49140
<host>
$ ./bin/lldb-dotest --platform-name remote-linux --platform-url connect://172.17.0.2:1234 --platform-working-dir /tmp/test_lldb --arch aarch64 -p TestSVE*
```
The SVE tests are the ones that found https://github.com/llvm/llvm-project/issues/97537.

And everything works as expected. I don't normally run remote tests with `ninja check-lldb`, so I just ran the above command in a couple of terminals at the same time and there weren't any problems.

```
\_ ./lldb-server platform --server --listen 0.0.0.0:1234 --gdbserver-port 49140
    \_ lldb-server platform --child-platform-fd 9 --gdbserver-port 49140
    \_ lldb-server platform --child-platform-fd 9 --gdbserver-port 49140
    \_ /home/davspi01/lldb-server gdbserver  --fd=9 --native-regs
    |   \_ sh -c /usr/bin/lsb_release -i
    |       \_ /usr/bin/python3 -Es /usr/bin/lsb_release -i
    \_ /home/davspi01/lldb-server gdbserver  --fd=9 --native-regs
        \_ sh -c /usr/bin/lsb_release -i
            \_ /usr/bin/python3 -Es /usr/bin/lsb_release -i
```
Snapshot of the processes during that, all makes sense to me. Main platform, two connected clients and each one has spawned a gdbserver.

slydiman commented:
> And everything works as expected. I don't normally run remote tests with `ninja check-lldb`, so I just ran the above command in a couple of terminals at the same time and there weren't any problems.

Note we are testing 
- API, Shell, unittests locally on Windows x86_64
- API, Shell, unittests locally on Linux x86_64
- API, Shell test cross Windows x86_64 host + Linux AArch64 target
- API, Shell test cross Linux x86_64 host + Linux AArch64 target

Currently we have 2 tests failed on Windows host/local. But these fails are not related to this patch.

slydiman commented:
@labath 
> Basically, the code should look similar to [this](https://github.com/llvm/llvm-project/blob/f71b63865140cf3c286baf3a77ba3e467f929504/lldb/source/Host/common/TCPSocket.cpp#L267), except that it will be listening to two unrelated sockets.

> Reasoning about a singlethreaded application is much easier than for a multithreaded one. We should already have all of the pieces to do this the right way, so let's just do that.

So, the TCPSocket class supports a list of listening native sockets. But Acceptor does not support them. MainLoop is not accessible anywhere.  I'm going to remove the new thread for accepting gdb connections and refactor Acceptor to listen on 2 ports in the same main thread instead.

slydiman commented:
I have updated the patch.
Note all changes in Socket.* and TCPSocket.cpp will go away after merging #106640.
The class Acceptor uses the MainLoop and contains the own list of listen sockets.
Please review. Thanks.

labath commented:
What I meant was something like [this](https://github.com/llvm/llvm-project/pull/106955). Basically, keep the socket code inside the socket class, but make it asynchronous. This way you can listen for as many sockets (and other things) as you like. It shouldn't be necessary to do the socket handling stuff externally.

I haven't looked into exactly how this would fit into the Acceptor class in lldb-server, but my preferred solutions would be:
- delete it (or at least significantly slim down it's functionality -- it's a very thin wrapper over the Socket class, and probably only exists due to historic reasons)
- make it callback-based as well

I don't like the idea of making Acceptor responsible for multiple ports for the same reason I did not want to do that for TCPSocket -- it's complicated and inflexible.

slydiman commented:
I have updated this patch based on #106955 to see how it will look and work.

slydiman commented:
Refactored after merging #106955.
Acceptor.cpp has been deleted.
Please review. Thanks.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> gdb_handles =
+        sock_gdb->Accept(*g_main_loop, [&platform, &gdb_inferior_arguments](
+                                           std::unique_ptr<TCPSocket> sock_up) {
+          printf("Gdb connection established.\n");
```
Do we want to print this for every new connection?

slydiman replied:
Removed.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
+            LLDB_LOGF(log, "gdbserver SharedSocket failed: %s",
+                      error.AsCString());
+
+          sock_up.reset();
```
the socket is going to get destroyed anyway.

slydiman replied:
Removed.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
+      g_main_loop->Run();
+    } while (!g_terminate);
+
+    gdb_handles->clear();
+    platform_handles->clear();
+
+    g_main_loop.reset();
+    sock_gdb.reset();
+    sock_platform.reset();
```
This isn't very c++-y. Could this be arranged so that the natural c++ destructor order destroys things in the right order. Perhaps by creating the mainloop inside this block (which means that the reaping callback would need to get it via an argument)?

slydiman replied:
I have added platform_tcp() and platform_named() and placed handles inside a block.

slydiman commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -150,12 +153,17 @@ static void client_handle(GDBRemoteCommunicationServerPlatform &platform,
+  if (g_single_pid != LLDB_INVALID_PROCESS_ID && g_single_pid == pid) {
+    // If not running as a server and the platform connection is closed
+    if (!g_terminate) {
+      g_terminate = true;
```
We need this callback only in case of non-server mode and tcp protocol.
Note this callback is called from the MonitorChildProcessThread. 
See the issue #101475. We must be very careful with this callback and we cannot change any arguments. 
I have removed g_terminate.

labath replied:
Yes, we need to be very careful, but I don't think this is how being careful looks like. `SetMonitorProcessCallback` takes a std::function, so we can add (or remove) as many arguments as we like.

labath replied:
Ok, I see what you meant with https://github.com/llvm/llvm-project/issues/101475 now. That said I don't think this solves anything, cause the same thing could now happen with the mainloop pointer (if it is destroyed just as this function is exiting). The way I'd ideally solve this is by ensuring that the main loop does not terminate (which means the main loop and the platform objects will not be destroyed) until this function is called.

slydiman replied:
Note this callback is called from an uncontrollable thread and any pending arguments in the std::function wrapper may be invalid during the process destroying procedure.
I have updated the logic around g_server and removed <atomic> and g_single_pid at all.

labath replied:
I'm not particularly happy with the design of this callback, but that doesn't mean it can't be used safely. The "uncontrollable" aspect means it is up to you (its user) to ensure the callback does not access any arguments after they are destroyed. It doesn't matter if they're local or global or how they're passed into the callback. The new implementation is better (because it's easier to reason about), but it still accesses the global main loop object, which will get destroyed some time after `main` returns. So, the question here is "is there anything ensuring that does not happen?" and not "is mainloop a global?". If there is (for example, because the main thread waits for the callback to call `RequestTermination()`), then I'm pretty this could be refactored to avoid the global. If there isn't, then this is a problem even though the variable is global.

slydiman replied:
There are a lot of discussions around `Detached threads accessing global or static objects` in the internet.
Detached threads are bad and the design of this callback is bad.
But if GDBRemoteCommunicationServerPlatform::g_spawned_pids is static (now) it reduces the chances of a crash to 0.(0)1%.
We have no handle of this thread. So the static global g_main_loop is the safest way and g_main_loop must be destroyed (almost) last. No 100% guarantee. But it is the better solution w/o redesigning the child process monitoring.
Ultimately we can just remove `fixes #101475` in the description because this patch improves the situation but do not fix it 100%.

labath replied:
I think I know the problems with detached threads pretty well (remember how I said I found bugs in lldb's dependencies -- [one of them](https://sourceware.org/bugzilla/show_bug.cgi?id=19951) is in glibc's pthread_detach). And I'm not interested in changing the design of the monitoring callback. It's detached and it's going to stay detached for a long time to come.

The thing I'm questioning is whether the code you are running inside the callback has to be detached (because the fact that the outer callback machinery is detached doesn't mean that the callback itself has to be detached).

The problems with detached threads are a runtime property. If a thread *accesses* something after it has been destroyed, you have a bug. I emphasised "access" because the access has to happen. Mere existence of a pointer/reference that would permit that access is not in itself a problem. So code like
```
int main() {
  std::condition_variable cv;
  std::thread([&]{cv.notify_one();}).detach();
  cv.wait();
  return;
}
```
is safe because even though the detached thread holds a reference to the local object (and may continue to hold it after its destroyed), the program ensures that the (only) usage of that reference happens while the object is alive. If I removed `cv.wait()` then we'd have a problem.

Why I'm mentioning this? Because this is exactly the pattern we have (or one that I think we should have) here, just instead of `cv.wait()` we have `mainloop.Run()` and `cv.notify()` is replaced by `mainloop.RequestTermination()`

If we can make sure things work this way, we're safe, or at least the (0.1%) problems are "not our problem". If we can't undetach this code, then we're exacerbating the problem, because even if someone wanted to make the undetach the monitoring callback machinery, he could make that work because the code inside it would be detached.

slydiman replied:
The child monitor callback is used in 2 places:

* lldb-platform.cpp: In case of non-server mode we need to exit after disconnecting (when the child process handling the platform connection exited). In theory we should be in the middle of MainLoop except the case when an assert happened somewhere. So access g_main_loop is safe here. But the safest way is to use a terminate flag and a dummy connection instead of accesing g_main_loop. We can pass a reference to MainLoop as a callback argument but it changes nothing. A global variable is much clear here.

* The child process handling the platform connection uses the single instance of GDBRemoteCommunicationServerPlatform, which uses a list of PIDs g_spawned_pids. This process will exit after disconnecting and the child process monitor callback is most dangerous point. Any processes may exit at any time. We need the PIDs list only for Handle_qKillSpawnedProcess response. It would be better to try to kill the process anyway and just check somehow that this PID is still alive. I think the best solution here is the static g_spawned_pids. Any other solutions are out the scope of this patch.

slydiman commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
+    sock_platform.reset();
+  } else {
+    Socket *conn_up = nullptr;
+    error = sock_named->Accept(conn_up);
```
I think it is out of scope of this patch.

slydiman replied:
Unix sockets accept only one socket. A single blocking accept() is enough. I think a MainLoop is redundant for unix sockets. Now the code in platform_named() is clear and consistent. Adding MainLoop and callbacks just will needlessly complicate the code.

labath replied:
I don't think so. The added code should be much smaller compared to the number of lines of code in lldb-platform that it will save. The impact is even bigger when you take into acount that lldb-platform is a huge messy monolith whereas the new function will be self-contained and unit-tested.
You don't have to make the existing Accept function go through the MainLoop like I did with TCP. You can just make an independent implementation if that's easier.

I'll also note that lldb-server currently supports accepting multiple connections over domain sockets (--server mode), whereas your implementation does not. if you added support for that, it would be even more duplicated code.

slydiman replied:
You have an idea of ​​what it should look like. I don't. 
I'm not expert in unix sockets. I have no idea how multiple named sockets may be connected simultaneously. 
I don't see how the following code may be reduced using a MainLoop and callbacks
```
  std::unique_ptr<Socket> sock_named = Socket::Create(protocol, false, error);
  error = sock_named->Listen(name, backlog);
  Socket *conn_up = nullptr;
  error = sock_named->Accept(conn_up);
  GDBRemoteCommunicationServerPlatform platform(protocol, 0);
  platform.SetConnection(std::unique_ptr<Connection>(new ConnectionFileDescriptor(conn_up)));
  client_handle(platform, args);
```
Anyway, this shouldn't be a stopper for this patch.

labath replied:
It will be reduced (to nothing) if it can use the same mainloop code that used by the tcp path. I think it is a showstopper because it makes the makes the patch very ugly. I can implement the mainloop/accept version for unix sockets, but it will only happen next week.

slydiman replied:
It will reduce the function platform_named() to nothing. But current platform_tcp() and GDBRemoteCommunicationServerPlatform will grow and will become terrible.
Unix sockets do not require ports and accept only 1 connection. GDBRemoteCommunicationServerPlatform has the different logic for TCP and unix sockets. We don't wait for a connection before launching gdbserver to provide fd in case of unix sockets. Using MainLoop for unix sockets will make lldb-platform.cpp and GDBRemoteCommunicationServerPlatform extremely confusing. Currently we don't need a separated child process to handle unix socket connection. Unix sockets are used only locally. The only point in using unix sockets - they are lite. But you want to make the connection process as hard as possible - unnecessary child processes, shared sockets for both lldb-platform and gdbserver, etc. I totally disagree. But if it is the showstopper, I will wait for your implementation of the mainloop/accept version for unix sockets.

slydiman commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
-    platform.SetConnection(std::unique_ptr<Connection>(conn));
+    do {
+      g_main_loop->Run();
+    } while (!g_terminate);
```
g_terminate has been removed.

slydiman commented on lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp:
```diff
@@ -667,7 +669,22 @@ ConnectionStatus
+  }
+  if (error_ptr)
+    *error_ptr = Status::FromErrorStringWithFormat(
+        "invalid file descriptor: \"%s\"", s.str().c_str());
```
ConnectionFileDescriptorPosix.cpp is unchanged now.

slydiman commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -160,66 +95,58 @@ GDBRemoteCommunicationServerPlatform::~GDBRemoteCommunicationServerPlatform() =
-  LLDB_LOGF(log, "Launching debugserver with: %s:%u...", hostname.c_str(),
-            *port);
+  LLDB_LOG(log, "Launching debugserver url='{0}', fd={1}...", url.str(),
+           (int64_t)fd);
```
Updated.

labath commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunication.cpp:
I think this file looks good now. In the interests of moving this forward, and reducing the size of this patch, could you split it into a separate patch?

slydiman replied:
I have created #107388. 
Note I did not include `#include "lldb/Host/Socket.h"` and shared_fd_t using because it requires changes in lldb/unittests/tools/lldb-server/tests/TestClient.*

labath commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.h:
```diff
@@ -25,90 +25,30 @@ namespace process_gdb_remote {
-  const std::string m_socket_scheme;
-  std::recursive_mutex m_spawned_pids_mutex;
-  std::set<lldb::pid_t> m_spawned_pids;
+  static std::mutex g_spawned_pids_mutex;
+  static std::set<lldb::pid_t> g_spawned_pids;
```
I guess this is related to #101475, but I think we should figure out a solution that does not require making these static.

slydiman replied:
There is no a safe way except static. 
Note GDBRemoteCommunicationServerPlatform works in a separated single thread process. I think it is ok.

labath commented on lldb/tools/lldb-server/lldb-gdbserver.cpp:
```diff
@@ -195,18 +195,31 @@ void ConnectToRemote(MainLoop &mainloop,
 
     // Create the connection.
-#if LLDB_ENABLE_POSIX && !defined _WIN32
+#if LLDB_ENABLE_POSIX
```
I don't think we actually need this. _WIN32 is basically the opposite of POSIX. AFAICT, https://reviews.llvm.org/D33213 was just being (too) defensive.

slydiman replied:
`!defined _WIN32` is redunded here because this block is inside the another #ifdef now:
```
#ifdef _WIN32
...
#else
 // no need to check !defined _WIN32 here
#endif
```

slydiman commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -160,66 +95,56 @@ GDBRemoteCommunicationServerPlatform::~GDBRemoteCommunicationServerPlatform() =
+      AddSpawnedProcess(pid);
+    LLDB_LOGF(log,
+              "GDBRemoteCommunicationServerPlatform::%s() "
+              "debugserver launched successfully as pid %" PRIu64,
```
Exactly. 
A non-error Status mean the process was launched. But don't add LLDB_INVALID_PROCESS_ID to the list. 
We can add assert() here but I don't like it.
```
  if (error.Success()) {
    pid = debugserver_launch_info.GetProcessID();
    if (pid != LLDB_INVALID_PROCESS_ID)
      AddSpawnedProcess(pid);
    LLDB_LOGF(log,
              "GDBRemoteCommunicationServerPlatform::%s() "
              "debugserver launched successfully as pid %" PRIu64,
              __FUNCTION__, pid);
  } else {
    LLDB_LOGF(log,
              "GDBRemoteCommunicationServerPlatform::%s() "
              "debugserver launch failed: %s",
              __FUNCTION__, error.AsCString());
  }
```

labath replied:
Why not? That's exactly what the assert is for. It says "it you say the process was launched, you better be able to give me its PID"

slydiman replied:
Added assert() to AddSpawnedProcess().

slydiman commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -160,66 +95,56 @@ GDBRemoteCommunicationServerPlatform::~GDBRemoteCommunicationServerPlatform() =
+  std::ostringstream url;
+  if (fd == SharedSocket::kInvalidFD) {
+    if (m_socket_protocol == Socket::ProtocolTcp) {
+      // Just check that GDBServer exists. GDBServer must be launched after
+      // accepting the connection.
+      if (!GetDebugserverPath(nullptr))
+        return Status::FromErrorString("unable to locate debugserver");
+      return Status();
```
It is inconvenient. It can be called from 
GDBRemoteCommunicationServerPlatform::Handle_qLaunchGDBServer()
and
lldb-platform.cpp, client_handle().
We can rename it to something like `CheckOrLaunchGDBServer()`.
BTW, Handle_qLaunchGDBServer() does not really launch the process now too in case of TCP protocol.

labath replied:
I am still looking for an answer to this one. If it's only the `Handle_qLaunchGDBServer` that needs to *not* launch an lldb-server then I don't see why the check couldn't/shouldn't be performed there.

slydiman commented:
I have separated changes `int connection_fd = -1` -> `shared_fd_t connection_fd = SharedSocket::kInvalidFD` to #107553 (NFC).

slydiman committed:
[lldb] Removed gdbserver ports map from lldb-server

Listen to gdbserver-port, accept the connection and run lldb-server gdbserver --fd on all platforms.
Parameters --min-gdbserver-port and --max-gdbserver-port are deprecated now.

This is the part 2 of #101283.

Fixes #97537, fixes #101475.

slydiman commented:
@labath
I tried to refactor this patch using #108188. No any profit. I had functions platform_tcp() and platform_named().
Now I can unify only 3 lines:
```
std::unique_ptr<Socket> sock_platform = Socket::Create(protocol, /*child_processes_inherit=*/false, error);
sock_platform->Listen(name, backlog);
...
main_loop->Run();
```
The rest 150 lines cannot be unified.
The code of platform_named() was clear and consistent
```
  std::unique_ptr<Socket> sock_named = Socket::Create(protocol, false, error);
  error = sock_named->Listen(name, backlog);
  Socket *conn_up = nullptr;
  error = sock_named->Accept(conn_up);
  GDBRemoteCommunicationServerPlatform platform(protocol, 0);
  platform.SetConnection(std::unique_ptr<Connection>(new ConnectionFileDescriptor(conn_up)));
  client_handle(platform, args);
```
Now it is really ugly because of a lot `if (protocol == Socket::ProtocolTcp)`, casts to TCPSocket, etc. 
It is necessary to resolve the listen ports for tcp.
We don't need the listen socket for gdb connection in case of ProtocolUnixDomain.
The accept callbacks are very different for unix and tcp sockets.
We used `GDBRemoteCommunicationServerPlatform platform(Socket::ProtocolTcp, gdbserver_port);` in the child process `lldb-server p --child-platform-fd`. If you want do execute such child process for unix socket too, it is necessary to transfer the protocol somehow too. It is an unnecessary complication.
I can commit the version where platform_named() uses `main_loop->Run();` but does not execute the child process `lldb-server p --child-platform-fd`. It will just add 30 lines. 
It is impossible to unify the code of functions platform_tcp() and platform_named().

slydiman commented:
@labath Let's decide how to finish this patch, please.
I would prefer to commit the new functionality. 
Any cosmetics and refactoring is possible at any time.

labath commented:
>  If you want to execute such child process for unix socket too, it is necessary to transfer the protocol somehow too.

I do want it. I don't think it's necessary to transfer the type of the socket, as that can be determined by inspecting the socket FD (`getsockopt(SO_DOMAIN/SO_PROTOCOL)`). I'd probably do it by creating a new `Socket::Create` static function which takes the FD, and instantiates the appropriate type based on what this function returns.

slydiman commented:
> I do want it. I don't think it's necessary to transfer the type of the socket, as that can be determined by inspecting the socket FD (getsockopt(SO_DOMAIN/SO_PROTOCOL)).

Oh, I expected this. We can determine non-tcp by missing or 0 gdb port. But it is impossible to separare ProtocolUnixDomain and ProtocolUnixAbstract. The difference is only the implementation of Socket::GetNameOffset().
getsockopt(SO_DOMAIN/SO_PROTOCOL) will not help.
We can provide `--listen same_host_port ` to the child process additionally to `--child-platform-fd X`  and move the parsing of the listen parameter to the beginning to get the correct protocol.
I will make the second commit with the version where I will mux functions platform_tcp() and platform_named() back together, complicated and hard as possible.

slydiman committed:
Refactored after #108188.

slydiman committed:
Used TCPSocket and DomainSocket constructors directly. Socket.cpp is untouched now.

slydiman commented:
I have updated Socket::Create() with minimal changes. But later I used TCPSocket and DomainSocket constructors directly to keep Socket.cpp untouched.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
+  error = sock_platform->Listen(
+      protocol == Socket::ProtocolTcp
+          ? llvm::formatv("{0}:{1}", hostname, platform_port).str()
+          : hostname,
```
Is `hostname` the unix socket path? That's definitely surprising. Can you instead define a single (string) variable called address (or whatever is the name of the argument to `Socket::Listen`, and assign it the appropriate value in the if block on line 415?

slydiman replied:
Done

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
+            *g_main_loop,
+            [progname, gdbserver_port, inferior_arguments, log_file,
+             log_channels](std::unique_ptr<Socket> sock_up) {
+              // If not running as a server, this process will not accept
```
AIUI, this is here to prevent multiple connections in non-server mode. Correct?

Is there any situation in non-server mode where we need to accept more than one successful connection?

slydiman replied:
We must accept only 1 platform connection in non-server mode. But we can accept multiple gdb connections.

labath replied:
Okay, in that case, I'd suggest that, instead of dropping connections on the floor, we close the listening handles (`platform_handles->clear()`) and also close the listening socket, so that the other side gets an ECONNREFUSED.

slydiman replied:
In case of unix sockets (no gdb listener) it will cause to immediately exit the main loop, exit the app and kill the child process which handles the platform connection.

labath replied:
I think that's actually great, as it allows us to synchronize with the exit of the child process and solve our global/detached thread problem -- after returning from MainLoop::Run, we just wait for a notification that the process has really exited. Something like:
```
std::promise<void> child_exited;
handles = socket.Accept(loop, [&]{
...
launch_info.SetMonitorProcessCallback([&child_exited]{child_exited.set_value();});
Launch(launch_info);
handles.clear();
});
loop.Run();
child_exited.get_future().get();
```

slydiman replied:
Done.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
-  if (error.Fail()) {
-    fprintf(stderr, "failed to create acceptor: %s", error.AsCString());
-    exit(socket_error);
+  std::string hostname;
+  uint16_t platform_port = 0;
+
+  // Try to match socket name as URL - e.g., tcp://localhost:5555
+  if (std::optional<URI> uri = URI::Parse(listen_host_port)) {
+    if (!Socket::FindProtocolByScheme(uri->scheme.str().c_str(), protocol)) {
+      fprintf(stderr, "Unknown protocol scheme \"%s\".",
+              uri->scheme.str().c_str());
+      return socket_error;
+    }
+    if (protocol == Socket::ProtocolTcp) {
+      hostname = uri->hostname;
+      if (uri->port) {
+        platform_port = *(uri->port);
+      }
+    } else
+      hostname = listen_host_port.substr(uri->scheme.size() + strlen("://"));
+  } else {
+    // Try to match socket name as $host:port - e.g., localhost:5555
+    llvm::Expected<Socket::HostAndPort> host_port =
+        Socket::DecodeHostAndPort(listen_host_port);
+    if (!llvm::errorToBool(host_port.takeError())) {
+      protocol = Socket::ProtocolTcp;
+      hostname = host_port->hostname;
+      platform_port = host_port->port;
+    } else
+      hostname = listen_host_port;
   }
```
Let's make a helper function for this. It could either return a decomposed/parsed socket address or maybe even a constructed socket object.

slydiman replied:
Done

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
-    client_handle(platform, inferior_arguments);
-  } while (g_server);
+    std::vector<MainLoopBase::ReadHandleUP> gdb_handles;
+    if (protocol == Socket::ProtocolTcp) {
```
Put this into a helper function (`AcceptGdbConnectionsIfNeeded`?). Also move the listening code there.

slydiman replied:
It is hard to combine listening and accepting code in the same helper. gdbserver_port may be updated after Listen() and it is used in the platform callback. I have created 2 helpers ListenGdbConnectionsIfNeeded() and AcceptGdbConnectionsIfNeeded().

slydiman committed:
Added parse_listen_host_port(), ListenGdbConnectionsIfNeeded() and AcceptGdbConnectionsIfNeeded().

slydiman commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
+#else
+      LLDB_LOGF(log,
+                "lldb-platform child: Unix domain sockets are not supported on "
+                "this platform.");
```
Probably you will never see that in stderr because it is a child process. But ok. Updated.

slydiman commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
+              "Ambiguous parameters --server --listen %s.\n"
+              "The protocol must be tcp for the server mode.",
+              listen_host_port.c_str());
+      return socket_error;
```
I don't know how it works. But ok. Removed.

slydiman committed:
Added std::promise<void> child_exited;

slydiman commented:
Polite ping.
Let's complite this PR, please.

labath commented:
I'm currently fighting some fires, so I didn't look at the new version yet, but I think the biggest blocker right now are the global variables in the platform class. I don't like globals, and I don't think they should be necessary to solve the problem at hand. If you can revert those (put them into a separate patch so we can continue the discussion), I think we can wrap this up relatively quickly.

slydiman commented:
> I think the biggest blocker right now are the global variables in the platform class. I don't like globals, and I don't think they should be necessary to solve the problem at hand. If you can revert those (put them into a separate patch so we can continue the discussion), I think we can wrap this up relatively quickly.

Ok, I will remove globals and let it crash.

slydiman committed:
Reverted m_spawned_pids.

slydiman commented:
> I don't like globals

Is this the only argument? I don't like globals too. But no one suggested better solution w/o huge refactoring of MonitorProcessCallback.

Updated. The child process will crash 50/50 at the end, but no one will see. The problem will remain hidden.

labath commented:
> > I don't like globals
> 
> Is this the only argument? 

No, it's not. You can find my second reason in the full text of my previous message. The third reason is that I think this is something that should be handled independenly of this (already larger that I'd like) patch.

> I don't like globals too. But no one suggested better solution w/o huge refactoring of MonitorProcessCallback.

You didn't ask for one either. Declaring it's [the only way](https://github.com/llvm/llvm-project/pull/104238#discussion_r1745282649) isn't the most traditional way of soliciting feedback. 

> 
> Updated. The child process will crash 50/50 at the end, but no one will see it. The problem will remain hidden. Removed `fixes #101475` in the description. I can create a separate patch after merging this PR.

:+1: 

I'm going to look at the patch now.

labath commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -318,75 +231,67 @@ GDBRemoteCommunicationServerPlatform::Handle_qKillSpawnedProcess(
+  // verify that we know anything about this pid.
+  if (SpawnedProcessFinished(pid)) {
+    // not a pid we know about
+    return SendErrorResponse(10); // ECHILD
```
Let's get rid of these comments. If we really wanted these to have a meaning, they should be one of the GDB_ESOMETHING constants defined in GDBRemoteCommunication.h.

slydiman replied:
Done.

labath commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -318,75 +231,67 @@ GDBRemoteCommunicationServerPlatform::Handle_qKillSpawnedProcess(
+  m_spawned_pids.insert(pid);
+}
+
+bool GDBRemoteCommunicationServerPlatform::SpawnedProcessFinished(
```
I'd suggest inverting this condition and calling this `SpawnedProcessIsRunning` (or sth like that). The reason being that "Finished" implies that process was running at some point, but this function does not distinguish a terminated process from one that was never run.

slydiman replied:
Done.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -119,20 +175,78 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+  return Status();
+}
+
+static Status AcceptGdbConnectionsIfNeeded(
```
return `Expected<handles>` here.

slydiman replied:
Done.

slydiman commented:
> I am still looking for an answer to this one. If it's only the Handle_qLaunchGDBServer that needs to not launch an lldb-server then I don't see why the check couldn't/shouldn't be performed there.

I cannot answer in the thread, no idea why. The answer is here
https://github.com/llvm/llvm-project/pull/104238#issuecomment-2368274458

LaunchGDBServer() may be called from the child process at the beginnig if inferior_arguments are present. The behavior depends on the protocol. gdbserver will be launched in case of unix sockets. Otherwise we need to check that gdbserver is exists and wait for the tcp gdb connection. This part cannot be optimized/simplified.

slydiman committed:
Cosmetics.

labath commented:
> > I am still looking for an answer to this one. If it's only the Handle_qLaunchGDBServer that needs to not launch an lldb-server then I don't see why the check couldn't/shouldn't be performed there.
> 
> I cannot answer in the thread, no idea why. The answer is here [#104238 (comment)](https://github.com/llvm/llvm-project/pull/104238#issuecomment-2368274458)
> 
> LaunchGDBServer() may be called from the child process at the beginnig if inferior_arguments are present. The behavior depends on the protocol. gdbserver will be launched in case of unix sockets. Otherwise we need to check that gdbserver is exists and wait for the tcp gdb connection. This part cannot be optimized/simplified.

Okay, that makes sense. Let's see how this goes.

slydiman commented:
#111033

DavidSpickett commented:
> ******************** TEST 'lldb-api :: lang/c/shared_lib_stripped_symbols/TestSharedLibStrippedSymbols.py' FAILED ********************

This test is flaky on Arm 32, I'll figure it out.

DavidSpickett commented:
Hi @slydiman, I am looking at commits that warrant a release note for LLDB 20, and I think your work on lldb-server's port map might fit the bill. If you want to do that yourself you can send a PR adding a bullet point to https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.md#changes-to-lldb, describing the changes.

Or you can tell me a 1/2 sentence summary of it and I'll meld it into release notes style. I know that it has helped my work because I only have to open 2 ports now (1 platform and 1 gdb-server), but I'm not 100% sure of the detail.

slydiman commented:
@DavidSpickett 
> Or you can tell me a 1/2 sentence summary of it and I'll meld it into release notes style

lldb-server now listens to a single port for gdbserver connections and provides it to the connection handler processes. This allows only 2 ports to be opened in the firewall. lldb-server now also works on Windows in the server mode.

Thanks

DavidSpickett commented:
Thanks! - https://github.com/llvm/llvm-project/commit/cfd7e024c6a97b0083f2e25a9d03d7dd516a0452#diff-9a29352a0c8c114f51b142b44cb1d5d6688787df1f78f3785ada73e1355a7379R463
